File: dev/test_data/cpp/test_main/aas_core_meta.v3/input/snippets/verification/data_specification_IEC_61360s_for_property_or_value_have_appropriate_data_type.cpp
bool DataSpecificationIec61360sForPropertyOrValueHaveAppropriateDataType(
  const std::vector<
    std::shared_ptr<types::IEmbeddedDataSpecification>
  >& embedded_data_specifications
) {
  for (const auto& embedded_data_specification : embedded_data_specifications) {
    const std::shared_ptr<
      types::IDataSpecificationContent
    >& content(
      embedded_data_specification->data_specification_content()
    );

	// NOTE (mristin):
	// We need to use dynamic cast due to virtual inheritance. Otherwise,
    // we would have used the model_type followed by a static cast.
	const types::IDataSpecificationIec61360* iec61360 = dynamic_cast<
	  const types::IDataSpecificationIec61360*
	>(content.get());

    if (iec61360 != nullptr) {
      const common::optional<
        types::DataTypeIec61360
      >& data_type(
        iec61360->data_type()
      );

      if (
        !data_type.has_value()
        || (
          constants::kDataTypeIec61360ForPropertyOrValue.find(*data_type)
          == constants::kDataTypeIec61360ForPropertyOrValue.end()
        )
      ) {
        return false;
      }
    }
  }

  return true;
}

File: dev/test_data/python/test_main/aas_core_meta.v3/input/snippets/Verification/data_specification_IEC_61360s_for_property_or_value_have_appropriate_data_type.py
def data_specification_iec_61360s_for_property_or_value_have_appropriate_data_type(
    embedded_data_specifications: Iterable[aas_types.EmbeddedDataSpecification],
) -> bool:
    """
    Check that :py:attr:`.types.DataSpecificationIEC61360.data_type`
    is defined appropriately for all data specifications whose content is given
    as IEC 61360.
    """
    for embedded_data_specification in embedded_data_specifications:
        if isinstance(
            embedded_data_specification.data_specification_content,
            aas_types.DataSpecificationIEC61360,
        ):
            iec61360 = embedded_data_specification.data_specification_content
            if (
                iec61360.data_type is None
                or iec61360.data_type
                not in aas_constants.DATA_TYPE_IEC_61360_FOR_PROPERTY_OR_VALUE
            ):
                return False

    return True
